### PR TITLE
realtek: support SFP+ slot on XGS1250-12.

### PR DESCRIPTION
**Pull request by TobleMiner**
This set of patches adds support for 1000Base-X and 10GBase-CR on the SFP+ port on the XGS1250-12.

Content of this patchset is inspired by the [rtl8214qf branch of @dangowrt's OpenWRT repo](https://github.com/dangowrt/openwrt/tree/rtl8214qf), porting parts forward to the current kernel and focussing on just SFP+ support to ensure the changes stay reviewable.

I've tested this patch on the SGX1250-12 with https://github.com/openwrt/openwrt/pull/14221 applied to fix QinQ issues. This PR does not depend on it though.

### CODE REVIEW

**Comment by Borromini** [comment]
@svanheule Another ping to take your mind off things :angel:

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
Comment (svanheule) [inline target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:831]: Were the dropped modes secretly not supported?

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (svanheule) [inline target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:1876]: `page` is uninitialized in this case

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (svanheule) [inline target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:1860]: Setting the defaults for `pre_en` and `post_en` again?

In any case this whole switch case has a weird mix of hex and decimal notation...

**Review by svanheule — CHANGES_REQUESTED** [review]
The second patch is really hard to review. Would it be possible to split this into changes that reorganize (e.g. the new `rtl9300_serdes_patch()` function) and changes that are functional? This would also make it easier to revert a change if required.

I've added notes on things that could really use some clarification. Since you taken changes from elsewhere, I'm not sure if you would be able to give names to some values either.

**Comment by Borromini** [comment]
Pending revision, I already tested this PR with my brand spanking new X520-DA1 card :nerd_face:. That just does 10 Gbit and 1 Gbit though, so can't test 5 or 2,5.

Logs look good:
```
Fri Jan  5 20:08:45 2024 kern.info kernel: [ 2253.384186] rtl83xx-switch switch@1b000000 lan12: Link is Down
Fri Jan  5 20:08:45 2024 kern.info kernel: [ 2253.391012] switch: port 12(lan12) entered disabled state
Fri Jan  5 20:08:45 2024 daemon.notice netifd: Network device 'lan12' link is down
Fri Jan  5 20:08:45 2024 kern.info kernel: [ 2253.399022] rtl83xx-switch switch@1b000000 lan12: Link is Up - 10Gbps/Full - flow control off
Fri Jan  5 20:08:45 2024 kern.info kernel: [ 2253.409637] switch: port 12(lan12) entered blocking state
Fri Jan  5 20:08:45 2024 kern.info kernel: [ 2253.415742] switch: port 12(lan12) entered forwarding state
Fri Jan  5 20:08:45 2024 daemon.notice netifd: Network device 'lan12' link is up
```

Ethtool output:

```
# ethtool lan12
Settings for lan12:
        Supported ports: [  ]
        Supported link modes:   1000baseX/Full
                                10000baseCR/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  1000baseX/Full
                                10000baseCR/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Speed: 10000Mb/s
        Duplex: Full
        Auto-negotiation: on
        Port: Direct Attach Copper
        PHYAD: 0
        Transceiver: internal
        Supports Wake-on: d
        Wake-on: d
        Link detected: yes
```

Ethtool on the client reports just this as advertised link mode but I suppose that's a client limitation?

`        Advertised link modes:  10000baseT/Full`

**Comment by davischw** [comment]
I tested the changes rebased on `main` (0fd3a6c6c6f079565f1f5de4709f3d4d71fc54c8) and `ethtool` (`ethtool-full`package) shows only `1000baseX/Full` as available link mode. When i connected port 12 with a SFP+ to RJ45 adapter (FiberStore SFP-10G-T) to port 10, the output of `ethtool` for that port shows a negotiated link speed of `10000Mb/s` with port 12 still showing `1000Mb/s`.

```
root@OpenWrt:/# ethtool lan12
Settings for lan12:
        Supported ports: [ MII ]
        Supported link modes:   1000baseX/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  1000baseX/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Full
        Auto-negotiation: on
        Port: MII
        PHYAD: 0
        Transceiver: internal
        Supports Wake-on: d
        Wake-on: d
        Link detected: no
```

```
root@OpenWrt:/# ethtool lan10
Settings for lan10:
        Supported ports: [  ]
        Supported link modes:   100baseT/Half 100baseT/Full
                                1000baseT/Full
                                10000baseT/Full
                                2500baseT/Full
                                5000baseT/Full
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  100baseT/Half 100baseT/Full
                                1000baseT/Full
                                10000baseT/Full
                                2500baseT/Full
                                5000baseT/Full
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  10baseT/Half 10baseT/Full
                                             100baseT/Half 100baseT/Full
                                             1000baseT/Half 1000baseT/Full
                                             10000baseT/Full
                                             2500baseT/Full
                                             5000baseT/Full
        Link partner advertised pause frame use: No
        Link partner advertised auto-negotiation: Yes
        Link partner advertised FEC modes: Not reported
        Speed: 10000Mb/s
        Duplex: Full
        Auto-negotiation: on
        Port: Twisted Pair
        PHYAD: 25
        Transceiver: external
        MDI-X: Unknown
        Supports Wake-on: d
        Wake-on: d
        Link detected: yes
```

```
root@OpenWrt:/# ethtool -m lan12
        Identifier                                : 0x03 (SFP)
        Extended identifier                       : 0x04 (GBIC/SFP defined by 2-wire interface ID)
        Connector                                 : 0x22 (RJ45)
        Transceiver codes                         : 0x10 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x1c
        Transceiver type                          : 10G Ethernet: 10G Base-SR
        Transceiver type                          : Extended: 10Gbase-T Short Reach
        Encoding                                  : 0x06 (64B/66B)
        BR, Nominal                               : 10300MBd
        Rate identifier                           : 0x00 (unspecified)
        Length (SMF,km)                           : 0km
        Length (SMF)                              : 0m
        Length (50um)                             : 0m
        Length (62.5um)                           : 0m
        Length (Copper)                           : 30m
        Length (OM3)                              : 0m
        Laser wavelength                          : 0nm
        Vendor name                               : FS
        Vendor OUI                                : 64:9d:99
        Vendor PN                                 : SFP-10G-T
        Vendor rev                                : A
        Option values                             : 0x00 0x1a
        Option                                    : RX_LOS implemented
        Option                                    : TX_FAULT implemented
        Option                                    : TX_DISABLE implemented
        BR margin, max                            : 0%
        BR margin, min                            : 0%
        Vendor SN                                 : F2230075431
        Date code                                 : 221114
```
```
root@OpenWrt:/# dmesg | grep rtl83xx-switch
[    2.239399] rtl83xx-switch switch@1b000000: configuring for fixed/internal link mode
[    2.258613] rtl83xx-switch switch@1b000000 lan1 (uninitialized): PHY [mdio-bus:00] driver [REALTEK RTL8218D] (irq=POLL)
[    2.272425] rtl83xx-switch switch@1b000000 lan2 (uninitialized): PHY [mdio-bus:01] driver [REALTEK RTL8218D] (irq=POLL)
[    2.286250] rtl83xx-switch switch@1b000000 lan3 (uninitialized): PHY [mdio-bus:02] driver [REALTEK RTL8218D] (irq=POLL)
[    2.300050] rtl83xx-switch switch@1b000000 lan4 (uninitialized): PHY [mdio-bus:03] driver [REALTEK RTL8218D] (irq=POLL)
[    2.313862] rtl83xx-switch switch@1b000000 lan5 (uninitialized): PHY [mdio-bus:04] driver [REALTEK RTL8218D] (irq=POLL)
[    2.327674] rtl83xx-switch switch@1b000000 lan6 (uninitialized): PHY [mdio-bus:05] driver [REALTEK RTL8218D] (irq=POLL)
[    2.341488] rtl83xx-switch switch@1b000000 lan7 (uninitialized): PHY [mdio-bus:06] driver [REALTEK RTL8218D] (irq=POLL)
[    2.355300] rtl83xx-switch switch@1b000000 lan8 (uninitialized): PHY [mdio-bus:07] driver [REALTEK RTL8218D] (irq=POLL)
[    2.398448] rtl83xx-switch switch@1b000000: Link is Up - 10Gbps/Full - flow control off
[    2.412573] rtl83xx-switch switch@1b000000 lan9 (uninitialized): PHY [mdio-bus:18] driver [Aquantia AQR113C] (irq=POLL)
[    2.446662] rtl83xx-switch switch@1b000000 lan10 (uninitialized): PHY [mdio-bus:19] driver [Aquantia AQR113C] (irq=POLL)
[    2.482925] rtl83xx-switch switch@1b000000 lan11 (uninitialized): PHY [mdio-bus:1a] driver [Aquantia AQR113C] (irq=POLL)
[    6.230524] rtl83xx-switch switch@1b000000 lan1: configuring for phy/xgmii link mode
[   22.446386] rtl83xx-switch switch@1b000000 lan1: configuring for phy/xgmii link mode
[   22.690845] rtl83xx-switch switch@1b000000 lan2: configuring for phy/xgmii link mode
[   22.839716] rtl83xx-switch switch@1b000000 lan3: configuring for phy/xgmii link mode
[   22.992077] rtl83xx-switch switch@1b000000 lan4: configuring for phy/xgmii link mode
[   23.136390] rtl83xx-switch switch@1b000000 lan5: configuring for phy/xgmii link mode
[   23.286328] rtl83xx-switch switch@1b000000 lan6: configuring for phy/xgmii link mode
[   23.421626] rtl83xx-switch switch@1b000000 lan7: configuring for phy/xgmii link mode
[   23.571462] rtl83xx-switch switch@1b000000 lan8: configuring for phy/xgmii link mode
[   23.801100] rtl83xx-switch switch@1b000000 lan9: configuring for phy/usxgmii link mode
[   23.994832] rtl83xx-switch switch@1b000000 lan10: configuring for phy/usxgmii link mode
[   24.115312] rtl83xx-switch switch@1b000000 lan11: configuring for phy/usxgmii link mode
[   24.239264] rtl83xx-switch switch@1b000000 lan12: configuring for inband/1000base-x link mode
[   25.528257] rtl83xx-switch switch@1b000000 lan1: Link is Up - 1Gbps/Full - flow control rx/tx
[   27.402726] rtl83xx-switch switch@1b000000 lan12: validation with support 0000000,00000000,00000000 failed: -22
[   29.808485] rtl83xx-switch switch@1b000000 lan10: Link is Up - 10Gbps/Full - flow control off
```

File: target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:831]: From what I can tell `sds_mode`  was set but then never used here. Threw it out since I was touching the code anyways to handle 1G modes. Am I missing something here?

Comment (svanheule) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:831]: OK, didn't realize it was dead code. Could split it off into a separate commit? Then it would be more clear why the code is being removed, and make it easier to find it back if someone ever want to resurrect it.

Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c:831]: Will do. Just giving this whole MR an overhaul since I finally found some more time for RTL39xx stuff

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:2891]: Unfortunately I do not know what exactly it does and why it works.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Fixed in updated patches.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Fixed in updated patches.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Both fixed in updated patches.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:2891]: Unfortunately I do not and any comment added would be based on my best guess only.

**Comment by TobleMiner** [comment]
Gave the changes a bit of an overhaul, fixing some ugly things here and there and hopefully making it easier to review. @svanheule could you have another look now, please? :slightly_smiling_face: 

 @davischw Unfortunately I do not have any 10GBaseT SFPs that work correctly with this switch, so I can not easily test what link modes are reported.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (svanheule) [inline target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: OK, the suspicious call goes away here.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (svanheule) [inline target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:3856]: No need to configure because it's done at a later point anyway?

**Review by svanheule — COMMENTED** [review]
Thanks for splitting the patches, now I feel like I have a better idea of the changes are trying to achieve. And thanks for the effort to clean up the sprawl of rtl930x serdes functions!

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Yep, this is just in the translation. This is not part of the final diff.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Think that is just in translation again and gets dropped in a later patch.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:2810]: I found this to give the cleanest diff, however I can of course replace the configure function by it instead. For readability I prefer it as is.

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: Yep

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (TobleMiner) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c:3856]: Yes. There is not really a point to configuration of serdes until we know what link mode we want. So it is now all deferred until we try to establish a link.

**Comment by TobleMiner** [comment]
@svanheule Are you happy with the PR as is or is there a strong wish for changing anything you have bought up in your second review (or even first review in case I missed something from there)?
I would like to get it merged/fully ready very soon so we do not go through another long loop with me having to put of the changes for quite some time due to other stuff keeping me busy :sweat_smile:

File: target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
Comment (svanheule) [reply target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c]: `sds_mode` in rtl9300_setup_serdes() is used in the same way as this one. So unless you really wanted to keep the `pr_info()` (I don't think it adds much value), you can drop these in the previous patch. That would make this patch and the next a bit shorter.

**Comment by svanheule** [comment]
The patches are mostly there, as far as I'm concerned. But while you're here, I want to see as much dead code go away as possible ;)
I hope you can find the time for those extra `sds_mode` cleanups, although I don't expect it should take very long. Sadly I'm not really set up to test any changes for 10G support. Anyway, at this point you understand this part of the code better than I do :)

**Comment by TobleMiner** [comment]
Ah! I totally overlooked that the phy_mode to sds_mode mapping is still duplicated in `rtl9300_serdes_setup`. Dropped that now.

**Review by svanheule — COMMENTED** [review]
This is always annoying as I run this before merging, but you still have some checkpatch warnings and errors:

[PATCH] realtek: 5.15: rtl93xx: remove unused SerDes mode selection
-:6: WARNING: Possible unwrapped commit description (prefer a maximum 75 chars per line)
total: 0 errors, 1 warnings, 42 lines checked

[PATCH] realtek: 5.15: rtl930x: introduce SerDes mode macros
-:63: WARNING: line length of 110 exceeds 100 columns
total: 0 errors, 1 warnings, 206 lines checked

[PATCH] realtek: 5.15: rtl93xx: add 1000Base-X and 10GBase-CR support on SerDes
-:115: ERROR: switch and case should be at the same indent
-:195: WARNING: braces {} are not necessary for single statement blocks
-:240: ERROR: code indent should use tabs where possible
-:241: ERROR: code indent should use tabs where possible
-:246: ERROR: code indent should use tabs where possible
-:247: ERROR: code indent should use tabs where possible
total: 5 errors, 1 warnings, 337 lines checked

**Comment by tristan-k** [comment]
Is the support for the sfp+ modules available in the `23.05.4` Release? I cant find any information in the [Release notes](https://openwrt.org/releases/23.05/notes-23.05.4).

**Comment by lucize** [comment]
@svanheule can't this be also backported to 23.05, it's a bit of pain for now on 6.6 with this chipset and stable seems to work with this, but it's missing by default

seems to recognize my DAC cable

```
[   48.665384] sfp sfp-p12: module CISCO-TYCO       1-2053783-2      rev S    sn TEDxxxADG1      dc 18051400
[   48.769323] rtl83xx-switch switch@1b000000 lan12: Link is Up - 10Gbps/Full - flow control off

```

also two types of modules


```
[  459.810900] sfp sfp-p12: module CISCO-FINISAR    FTLX8571D3BCL-C2 rev A    sn FNSxxxxWXN      dc 140705
[  459.821732] rtl93xx_phylink_mac_config port 27, mode 2, phy-mode: 10gbase-r, speed -1, link 0
[  553.041664] rtl83xx-switch switch@1b000000 lan12: Link is Up - 10Gbps/Full - flow control off
[  553.052862] rtl83xx-switch switch@1b000000 lan12: Link is Down
[  555.719661] sfp sfp-p12: module removed
[  562.613611] sfp sfp-p12: module FINISAR CORP.    FTLX8571D3BCL-QL rev A    sn APL0DBH          dc 130517
[  562.715592] rtl83xx-switch switch@1b000000 lan12: Link is Up - 10Gbps/Full - flow control off

```

**Comment by lucize** [comment]
sure, just that 24.10 it's with non-working 6.6 (from what I saw in the forum) but just wanted to see if will be backported

**Comment by leo5593** [comment]
Hello,

I have a XGS1210-12, I have done the support with the 23.05 version : 
https://github.com/leo5593/openwrt/tree/xgs1210-12_support

I send some time to try to made the SFP+ working and 1 line of code was missing (SDS register write), when added 10G SFP+ are working.
if this can help.

I will check the 24.10 with the 6.6 kernel to be more up to date.

**Comment by Neustradamus** [comment]
Dear all, @leo5593 can help, can you see with?

**Comment by leo5593** [comment]
Hello,

I have checked the 6.6 kernel but there is only the SFP working some how.

So going back to 5.15 kernel and back port some code from 6.6 but that looks pretty similar to the commit on 5.15 from Feb 2024:
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=9daf4dff6bd2908040a281efc082ac09b3d25871

This has provide 1G SFP to be working.

After a while looking for proper vlan 802.1Q proper management, the issue was on VLAN_PORT_FWD_CTRL and I reverted to an old change on "rtl930x_vlan_fwd_on_inner" function and vlan are working properly with "bridge vlan".

This give a good start for the support of the XGS1210-12.

I have some SFP and SFP+ that are "not supported" but are working with the OEM firmware.

`
[ 3586.896323] sfp sfp-p12: module FS               SFP-GB-GE-T      rev F    sn F2032218720 
[ 3586.986770] In rtl93xx_phylink_validate port 27, state is 0 ()
[ 3586.986801] rtl93xx_phylink_validate leaving supported: 0000000,00000000,00000000
[ 3586.993384] rtl83xx-switch switch@1b000000 lan12: validation with support 0000000,00000000,00000000 failed: -22
[ 3587.013128] sfp sfp-p12: sfp_add_phy failed: -22
root@OpenWrt:/etc/config# ethtool -m lan12
        Identifier                                : 0x03 (SFP)
        Extended identifier                       : 0x04 (GBIC/SFP defined by 2-wire interface ID)
        Connector                                 : 0x22 (RJ45)
        Transceiver codes                         : 0x00 0x00 0x00 0x08 0x00 0x00 0x00 0x00 0x00
        Transceiver type                          : Ethernet: 1000BASE-T
        Encoding                                  : 0x01 (8B/10B)
        BR, Nominal                               : 1300MBd
        Rate identifier                           : 0x00 (unspecified)
        Length (SMF,km)                           : 0km
        Length (SMF)                              : 0m
        Length (50um)                             : 0m
        Length (62.5um)                           : 0m
        Length (Copper)                           : 100m
        Length (OM3)                              : 0m
        Laser wavelength                          : 0nm
        Vendor name                               : FS
        Vendor OUI                                : 00:00:00
        Vendor PN                                 : SFP-GB-GE-T
        Vendor rev                                : F
        Option values                             : 0x00 0x10
        Option                                    : TX_DISABLE implemented
        BR margin, max                            : 0%
        BR margin, min                            : 0%
        Vendor SN                                 : F2032218720
        Date code                                 : 211215
`

I will keep looking on this side.

Commits are at https://github.com/leo5593/openwrt/commits/xgs1210-12_support/

**Pull request closed without merging**
